src/map/entities/charentity.cpp

```diff
@@ -1077,6 +1077,80 @@ void CCharEntity::OnCastFinished(CMagicState& state, action_t& action)
 
                 StatusEffectContainer->DelStatusEffectSilent(EFFECT_CHAIN_AFFINITY);
             }
+
+            if (actionTarget.param > 0 &&
+                PSpell->dealsDamage() &&
+                PSpell->getSpellGroup() == SPELLGROUP_BLACK &&
+                (StatusEffectContainer->HasStatusEffect(EFFECT_IMMANENCE)))
+            {
+                SUBEFFECT effect = SUBEFFECT_NONE;
+                switch (PSpell->getSpellFamily())
+                {
+                    case SPELLFAMILY_STONE:
+                        effect = battleutils::GetSkillChainEffect(PTarget, 4, 0, 0);
```

Comment:
Does `PSpell->getElement()` match the element that would be set in GetSkillChainEffect? If it does, you could have a less complicated case/switch with one case for spells that _can_ skillchain and a default case for ones that can't.

I don't think Bio/Drain can make a skillchain here, so checking spellfamily is probably required.